docs/docs/how_to/how-to-recursion.md

```diff
@@ -0,0 +1,178 @@
+
+:::info
+
+The `proofAsFields` has a constant size `[Field; 93]`. However, currently the backend doesn't remove the public inputs from the proof when converting it.
+
+This means that if your `main` circuit has two public inputs, then you should also modify the recursive circuit to accept a proof with the public inputs appended. This means that in our example, since `y` is a public input, our `proofAsFields` is of type `[Field; 94]`.
+
+Verification keys in Barretenberg are always of size 114.
+
+:::
+
+:::warning
+
+One common mistake is to forget *who* makes this call.
+
+In a situation where Alice is generating the `main` proof, if she generates the proof artifacts and sends them to Bob, which gladly takes them as true, this would mean Alice could prove anything!
+
+Instead, Bob needs to make sure *he* extracts the proof artifacts, using his own instance of the `main` circuit backend. This way, Alice has to provide a valid proof for the correct `main` circuit.
+
+:::
+
+## Step 4 - Recursive proof generation
+
+With the artifacts, generating a recursive proof is no different from a normal proof. You simply use the `backend` (with the recursive circuit) to generate it:
+
+```js
+const recursiveInputs = {
+    verification_key: vkAsFields, // array of length 114
+    proof: proofAsFields, // array of length 93 + size of public inputs
+    publicInputs: [mainInput.y], // using the example above, where `y` is the only public input
+    key_hash: vkHash,
+    input_aggregation_object: Array(16).fill(0) // this circuit is verifying a non-recursive proof, so there's no input aggregation object: just use zero
+}
+
+const { witness, returnValue } = noir.execute(recursiveInputs) // we're executing the recursive circuit now!
+const { proof, publicInputs } = backend.generateFinalProof(witness)
+const verified = backend.verifyFinalProof({ proof, publicInputs })
+```
+
+You can obviously chain this proof into another proof. In fact, if you're using recursive proofs, you're probably interested of using them this way! In that case, you should keep in mind the `returnValue`, as it will contain the `input_aggregation_object` for the next proof.
+
+:::tip
+
+Managing circuits and "who does what" can be confusing. To make sure your naming is consistent, you can keep them in an object. For example:
+
+```js
+const circuits = {
+main: mainJSON, 
+recursive: recursiveJSON
+}
+const backends = {
+main: new BarretenbergBackend(circuits.main),
+recursive: new BarretenbergBackend(circuits.recursive)
+}
+const noirs = {
+main: new Noir(circuits.main, backends.main),
+recursive: new Noir(circuits.recursive, backends.recursive)
+}
+```
+
+This allows you to neatly call exactly the method you want without conflicting names:
+
+```js
+// Alice runs this 👇
+const { witness: mainWitness } = await noirs.main.execute(input)
+const proof = await backends.main.generateIntermediateProof(witness)
+
+// Bob runs this 👇
```

Comment:
Bob should also run `const verified = await backend.verifyIntermediateProof({ proof, publicInputs })`, correct? If Alice provides him a bad proof, it doesn't make sense for to create additional proofs using it as an input because they'll fail, right?

Reply:
Yeah that's right, in a real use-case scenario it would be wasteful to try to prove the verification of something if it fails to do so in plain.

I'll update 👍